indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java

```diff
@@ -0,0 +1,1958 @@
+                        row,
+                        sequenceToUse.getSequenceName(),
+                        committerSupplier,
+                        // true for kafka, false for kinesis
+                        isSkipSegmentLineageCheck,
+                        // do not allow incremental persists to happen until all the rows from this batch
+                        // of rows are indexed
+                        false
+                    );
+
+                    if (addResult.isOk()) {
+                      // If the number of rows in the segment exceeds the threshold after adding a row,
+                      // move the segment out from the active segments of BaseAppenderatorDriver to make a new segment.
+                      if (addResult.isPushRequired(tuningConfig) && !sequenceToUse.isCheckpointed()) {
+                        sequenceToCheckpoint = sequenceToUse;
+                      }
+                      isPersistRequired |= addResult.isPersistRequired();
+                    } else {
+                      // Failure to allocate segment puts determinism at risk, bail out to be safe.
+                      // May want configurable behavior here at some point.
+                      // If we allow continuing, then consider blacklisting the interval for a while to avoid constant checks.
+                      throw new ISE("Could not allocate segment for row with timestamp[%s]", row.getTimestamp());
+                    }
+
+                    if (addResult.getParseException() != null) {
+                      handleParseException(addResult.getParseException(), record);
+                    } else {
+                      rowIngestionMeters.incrementProcessed();
+                    }
+                  } else {
+                    rowIngestionMeters.incrementThrownAway();
+                  }
+                }
+                if (isPersistRequired) {
+                  Futures.addCallback(
+                      driver.persistAsync(committerSupplier.get()),
+                      new FutureCallback<Object>()
+                      {
+                        @Override
+                        public void onSuccess(@Nullable Object result)
+                        {
+                          log.info("Persist completed with metadata [%s]", result);
+                        }
+
+                        @Override
+                        public void onFailure(@ParametersAreNonnullByDefault Throwable t)
+                        {
+                          log.error("Persist failed, dying");
+                          backgroundThreadException = t;
+                        }
+                      }
+                  );
+                }
+              }
+              catch (ParseException e) {
+                handleParseException(e, record);
+              }
+
+
+              currOffsets.put(
+                  record.getPartitionId(),
+                  getNextSequenceNumber(
```

Comment:
If I'm reading this right,  calling `getNextSequenceNumber() here is super expensive for Kinesis since it makes another HTTP request on every message read. Do we need to do this? Can we track the last sequence number processed instead of the next one (which is how it was done in the original impl to avoid making this call)?